python/datafusion/udf.py

```diff
@@ -246,3 +246,229 @@ def udaf(
             state_type=state_type,
             volatility=volatility,
         )
+
+
+class WindowEvaluator(metaclass=ABCMeta):
+    """Evaluator class for user defined window functions (UDWF).
+
+    It is up to the user to decide which evaluate function is appropriate.
+
+    |``uses_window_frame``|``supports_bounded_execution``|``include_rank``|function_to_implement|
+    |---|---|----|----|
+    |False (default)      |False (default)               |False (default) | ``evaluate_all``           |
+    |False                |True                          |False           | ``evaluate``               |
+    |False                |True/False                    |True            | ``evaluate_all_with_rank`` |
+    |True                 |True/False                    |True/False      | ``evaluate``               |
+    """  # noqa: W505
+
+    def memoize(self) -> None:
+        """Perform a memoize operation to improve performance.
+
+        When the window frame has a fixed beginning (e.g UNBOUNDED
```

Comment:
Q for my own understanding:

Should this be "When the window frame has a fixed **bound** (beginning or end)..."?

I don't see how `last_value` could benefit from `memoize` because of a fixed bound?


(If I'm correct, I'll go fix the docs upstream)

https://docs.rs/datafusion/latest/datafusion/logical_expr/trait.PartitionEvaluator.html#method.memoize

Reply:
You're right! `last_value` does not memoize the results, as evidenced here: https://github.com/apache/datafusion/blob/main/datafusion/physical-expr/src/window/nth_value.rs#L200